### DISCUSSION TITLE
Electron's native Node

### DISCUSSION THREAD

**Discussion by ccheraa** [eclipse-theia/theia#8449] [General] [answered]
Recently, I've come across an error that occurs when I try to start electron-app:
```
Error: The module '/path/to/native/module.node'
was compiled against a different Node.js version using
NODE_MODULE_VERSION $XYZ. This version of Node.js requires
NODE_MODULE_VERSION $ABC. Please try re-compiling or re-installing
the module (for instance, using `npm rebuild` or `npm install`).
```
this error is related to "Electron's native Node" and is mentioned on [Electron's documentation](https://www.electronjs.org/docs/tutorial/using-native-node-modules#using-native-node-modules)

Currently, I'm fixing this by running `electron-rebuild`, as suggested by Electron's documentation, but I'm wondering if anyone had this problem before, and if there's a better approach to fix it

**Comment by vince-fugnitto** [comment]
@ccheraa you must rebuild the native modules when starting `electron`.

We have the commands (for easy-use):

- `yarn start:browser`: rebuilds for the `browser` target and starts the example browser app
- `yarn start:electron`: rebuilds for the `electron` target and starts the example electron app

What was your use-case which resulted in the error?

**Comment by vince-fugnitto** [comment, marked as answer]
@ccheraa I’m sorry but what is the question? (I see you unmarked the answer without any justification or follow-up). The native modules must be rebuilt for electron when switching between browser and electron targets. As you mentioned previously, this is officially documented. I’m not sure what the actual question is?

**Comment by ccheraa** [comment]
Actually, `yarn start:electron` didn't fix my problem.

this is the error I get:
```
Error: The module '/home/dc/work/fusion-studio-extension/node_modules/drivelist/build/Release/drivelist.node'
was compiled against a different Node.js version using
NODE_MODULE_VERSION 64. This version of Node.js requires
NODE_MODULE_VERSION 80. Please try re-compiling or re-installing
the module (for instance, using `npm rebuild` or `npm install`).
```
I had to run `yarn workspace electron-app run electron-rebuild` for the Electron app to work.
But after that, the browser app doesn't work, and that's the error:
```
Error: The module '/home/dc/work/fusion-studio-extension/node_modules/drivelist/build/Release/drivelist.node'
was compiled against a different Node.js version using
NODE_MODULE_VERSION 80. This version of Node.js requires
NODE_MODULE_VERSION 64. Please try re-compiling or re-installing
the module (for instance, using `npm rebuild` or `npm install`).
```
and to fix it I had to reinstall the `drivelist` package

Here are the steps to recreate the problem:
```bash
git clone https://github.com/ccheraa/fusion-studio-extension
git checkout electron
yarn
cd electron-app
yarn start # works, because electron-rebuild is run in postinstall script
cd ../browser-app
yarn start # fails
```

*edited:* I'm sorry my comment didn't go through this morning, internet connection issue @vince-fugnitto

**Reply by vince-fugnitto** [reply]
@ccheraa no problem, rebuilding `drivelist` has been fixed as of yesterday:
- https://github.com/eclipse-theia/theia/pull/8454

As mentioned, if you are switching between `browser` and `electron` targets the native modules must be rebuilt.

```
git clone https://github.com/ccheraa/fusion-studio-extension
git checkout electron
yarn
cd electron-app
yarn start # works, because electron-rebuild is run in postinstall script
cd ../browser-app
yarn start # fails
```

This case will not work unless you rebuild the native modules (using `rebuild:browser` or `rebuild:electron`) before starting the respective targets.

**Reply by ccheraa** [reply]
Sorry, I didn't include it in my code, but even using `rebuild:browser` and `rebuild:electron` didn't fix the problem.
But now, I've seen your commit from yesterday, and that is the fix for my problem.
But I noticed that the change isn't up on NPMJS yet, do you have an idea on how/when will the npm package be updated?

**Reply by vince-fugnitto** [reply]
> Sorry, I didn't include it in my code, but even using `rebuild:browser` and `rebuild:electron` didn't fix the problem.
> But now, I've seen your commit from yesterday, and that is the fix for my problem.
> But I noticed that the change isn't up on NPMJS yet, do you have an idea on how/when will the npm package be updated?

The change is included in the `next` version (or the specific version, ex: `1.6.0-next.d7605539`), and will be released as part of the official `1.16.0` release at the end of the month.
